net: sockets: Fix datagram AF_PACKET socket handling

rlubos opened the pull request:
This PR fixes the datagram AF_PACKET socket handling, which improperly received packet a packet before it was processed by L2 layer.

Additionally, fix the existing unit test to work correctly after update, so that the packets used in the test are not rejected by L2. Finally add two new unit tests for DGRAM sockets to improve test coverage.

Fixes #45514

rlubos committed:
net: context: Improve packet socket interface selection on sendto()

Currently, the packet socket implementation in net_context required that
netowrk interface to transmit the packet to was set on every sendto()
call. This spoils the whole idea of binding a socket. Fix this, by
checking first if the net_conext is already bound to a particular
interface, and if so, do not throw an error in case sll_ifindex field is
not a valid interface.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: pkt: Add flag indicating if pkt has already been processed by L2

Add a new flag, indicating if packet has already been processed by L2 or
not. This can be used with AF_PACKET sockets, to select a proper socket
type for currently processed packet.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: conn: Add proper processing of DGRAM packet sockets

Datagram AF_PACKET sockets were not processed properly by the net stack.
Instead of receving a packet already processed L2, and thus with L2
header trimmed, it was receiving a raw, unprocessed packet.

Fix this by calling net_packet_socket_input() for the second time, after
L2 has processed the packet. An updated connection handler module will
forward the packet correctly based on the corresponding socket type and
packet L2 processing status.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
tests: net: sockets: Update packet socket tests

Fix the existing DGRAM packet socket test. The test will now specify
LL destination address and protocol type. Also, update the fake ethernet
device send handler, so that it sets the packet network interface
corretly, so that the packet is not rejected at Ethernet L2.

Additionally, add two additional test cases verifying datagram packet
sockets. Fist test will exchange packet between RAW and DGRAM socket,
making sure that net stack forwards/removes L2 header correctly. The
second test verifies that it's possible to recieve a packet on both RAW
and DGRAM socket, making sure the received data has correct format.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: sockets: packet: Remove obsolete TODO

It turns out that all of the items listed under TODO section for packet
socket implmeentation has been adressed over time, threfore remove the
obsolete TODO.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos commented:
CC @krish2718

krish2718 commented:
LGTM, let me test and update.

jukkar reviewed (dismissed):
Thanks Robert, LGTM

krish2718 commented:
I have ported this to ncs-zephyr (include/zephyr/net -> include/net) and tested with my app, the L2 packets are being dropped at `subsys/net/l2/ethernet/ethernet.c: ethernet_recv` because the type is not supported, my app is using a custom protocol for L2, so, I guess it needs to be patched as well?

krish2718 commented:
In [wpa_supplicant l2 packet ](https://github.com/nrfconnect/sdk-hostap/pull/1/files#diff-0f1819c18f80c5e9edc42583b253643fa0c393100d1414117d9ee1d8109933d2R153) an arbitrary protocol will be passed, in this specific case its EAPOL 0x888e.

rlubos commented:
> In [wpa_supplicant l2 packet ](https://github.com/nrfconnect/sdk-hostap/pull/1/files#diff-0f1819c18f80c5e9edc42583b253643fa0c393100d1414117d9ee1d8109933d2R153) an arbitrary protocol will be passed, in this specific case its EAPOL 0x888e.

Ok, in such case we need to handle EAPoL in `ethernet_recv()` specifically, or maybe we could change the default behaviour in case DGRAM packet sockets are enabled, i. e., pass unrecognized EtherTypes as well to the application in such case. @jukkar Do you have an opinion on that?

krish2718 commented:
Yes, at Ethernet L2 level allow all packets, and then filter them at packet socket level based on protocol registered during socket open.

carlescufi commented:
@krish2718 can you please approve if it fixes your issue?

krish2718 commented:
> @krish2718 can you please approve if it fixes your issue?

No, it doesn't fix my issue.

Before fix: packets received without L2 processsing
With fix: packets dropped during L2 processing, so, not received by app.

rlubos committed:
net: ethernet: Add Kconfig option to forward unrecognized EtherType frames

Add Kconfig option to the Ethernet L2 which allows to forward frames
with unknown EtherType further into the stack. This can be useful for
packet sockets, where further frame processing is application dependent.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos commented:
> > @krish2718 can you please approve if it fixes your issue?
> 
> No, it doesn't fix my issue.
> 
> Before fix: packets received without L2 processsing 
> With fix: packets dropped during L2 processing, so, not received by app.

I think that a broken datagram socket processing and unsupported EtherType are two separate topics (especially that there were no mention of it in the original report), but I have added an extra commit here which allows to forward any Ethernet frame type to the application. Please give it a try.

krish2718 commented:
> > > @krish2718 can you please approve if it fixes your issue?
> > 
> > 
> > No, it doesn't fix my issue.
> > Before fix: packets received without L2 processsing
> > With fix: packets dropped during L2 processing, so, not received by app.
> 
> I think that a broken datagram socket processing and unsupported EtherType are two separate topics (especially that there were no mention of it in the original report), but I have added an extra commit here which allows to forward any Ethernet frame type to the application. Please give it a try.

The packets are coming now, but they are being treated as RAW packets (proto = ETH_P_ALL: through conn_raw_socket) and hence no L2 processing is done, L2 header is still intact. You can easily test this by modifying the testcase to set protocol as 0x888e.

krish2718 commented:
Apologies, I have picked up wrong ZEPHYR that didn't have your full changes, it works now.

rlubos commented:
> Apologies, I have picked up wrong ZEPHYR that didn't have your full changes, it works now. Feel free to merge.

CC @carlescufi

jukkar commented on subsys/net/l2/ethernet/ethernet.c:
```diff
@@ -254,8 +254,13 @@ static enum net_verdict ethernet_recv(struct net_if *iface,
 		goto drop;
 #endif
 	default:
+#if defined(CONFIG_NET_ETHERNET_FORWARD_UNRECOGNISED_ETHERTYPE)
```
Minor nit, we could actually do `if (IS_ENABLED(....))` here.

rlubos replied:
Right, fixed.

krish2718 commented:
I will close my PR https://github.com/zephyrproject-rtos/zephyr/pull/45513 as that test is covered here.
